`test-runtime` failure is legit:

```
[2023-04-07T13:30:07.260Z] 	 [13:26:19] FAIL: manager_privileged_test.go:149: EgressGatewayTestSuite.TestEgressGatewayManager
[2023-04-07T13:30:07.260Z] 	 [13:26:19] 
[2023-04-07T13:30:07.260Z] 	 [13:26:19] manager_privileged_test.go:246:
[2023-04-07T13:30:07.260Z] 	 [13:26:19]     assertEgressRules(c, []egressRule{
[2023-04-07T13:30:07.260Z] 	 [13:26:19]         {ep1IP, destCIDR, egressIP1, node1IP},
[2023-04-07T13:30:07.260Z] 	 [13:26:19]         {ep1IP, excludedCIDR1, egressIP1, zeroIP4},
[2023-04-07T13:30:07.260Z] 	 [13:26:19]         {ep2IP, destCIDR, zeroIP4, node2IP},
[2023-04-07T13:30:07.260Z] 	 [13:26:19]     })
[2023-04-07T13:30:07.260Z] 	 [13:26:19] manager_privileged_test.go:609:
[2023-04-07T13:30:07.260Z] 	 [13:26:19]     c.Assert(policyVal.GetGatewayIP().Equal(r.gatewayIP), Equals, true)
[2023-04-07T13:30:07.260Z] 	 [13:26:19] ... obtained bool = false
[2023-04-07T13:30:07.260Z] 	 [13:26:19] ... expected bool = true
[2023-04-07T13:30:07.260Z] 	 [13:26:19] 
[2023-04-07T13:30:07.260Z] 	 [13:26:19] OOPS: 0 passed, 1 FAILED
[2023-04-07T13:30:07.260Z] 	 [13:26:19] --- FAIL: Test (0.08s)
[2023-04-07T13:30:07.260Z] 	 [13:26:19] FAIL
[2023-04-07T13:30:07.260Z] 	 [13:26:19] 	github.com/cilium/cilium/pkg/egressgateway	coverage: 68.2% of statements
[2023-04-07T13:30:07.260Z] 	 [13:26:19] FAIL	github.com/cilium/cilium/pkg/egressgateway	0.105s
[2023-04-07T13:30:07.260Z] 	 [13:26:19] ?   	github.com/cilium/cilium/pkg/endpoint/regeneration	[no test files]
```

we forgot to bring the egressgw manager unit tests up to date with the new gatewayIP format

--- a/bpf/lib/egress_policies.h
+++ b/bpf/lib/egress_policies.h
@@ -16,6 +16,11 @@
 #define EGRESS_STATIC_PREFIX (sizeof(__be32) * 8)
 #define EGRESS_PREFIX_LEN(PREFIX) (EGRESS_STATIC_PREFIX + (PREFIX))
 #define EGRESS_IPV4_PREFIX EGRESS_PREFIX_LEN(32)
+/* These are special IP values in the CIDR 0.0.0.0/8 range that map to specific
+ * case for in the egress gateway policies handling.
+ */
+#define EGRESS_GATEWAY_NO_GATEWAY (0)
+#define EGRESS_GATEWAY_EXCLUDED_CIDR bpf_htonl(1)
 
 static __always_inline
 struct egress_gw_policy_entry *lookup_ip4_egress_gw_policy(__be32 saddr, __be32 daddr)
@@ -38,10 +43,9 @@ bool egress_gw_request_needs_redirect(struct iphdr *ip4, __u32 *tunnel_endpoint)
 	if (!egress_gw_policy)
 		return false;
 
-	/* If the gateway IP is 0.0.0.0 it means this is an excluded CIDR, so
-	 * skip redirection
-	 */
-	if (!egress_gw_policy->gateway_ip)
+	/* FIXME: no gateway traffic should be dropped in a future release */
+	if (egress_gw_policy->gateway_ip == EGRESS_GATEWAY_NO_GATEWAY ||
+	    egress_gw_policy->gateway_ip == EGRESS_GATEWAY_EXCLUDED_CIDR)
 		return false;
 
 	/* If the gateway node is the local node, then just let the
@@ -65,10 +69,8 @@ bool egress_gw_snat_needed(struct iphdr *ip4, __be32 *snat_addr)
 	if (!egress_gw_policy)
 		return false;
 
-	/* If the gateway IP is 0.0.0.0 it means this is an excluded CIDR, so
-	 * skip SNAT
-	 */
-	if (!egress_gw_policy->gateway_ip)
+	if (egress_gw_policy->gateway_ip == EGRESS_GATEWAY_NO_GATEWAY ||
+	    egress_gw_policy->gateway_ip == EGRESS_GATEWAY_EXCLUDED_CIDR)
 		return false;
 
 	*snat_addr = egress_gw_policy->egress_ip;
@@ -87,10 +89,8 @@ bool egress_gw_reply_needs_redirect(struct iphdr *ip4, __u32 *tunnel_endpoint,
 	if (!egress_policy)
 		return false;
 
-	/* If the gateway IP is 0.0.0.0 it means this is an excluded CIDR, so
-	 * skip reply redirect
-	 */
-	if (!egress_policy->gateway_ip)
+	if (egress_policy->gateway_ip == EGRESS_GATEWAY_NO_GATEWAY ||
+	    egress_policy->gateway_ip == EGRESS_GATEWAY_EXCLUDED_CIDR)
 		return false;
 
 	info = ipcache_lookup4(&IPCACHE_MAP, ip4->daddr, V4_CACHE_KEY_LEN, 0);

--- a/bpf/tests/tc_egressgw_redirect.c
+++ b/bpf/tests/tc_egressgw_redirect.c
@@ -219,7 +219,7 @@ int egressgw_skip_excluded_cidr_redirect_setup(struct __ctx_buff *ctx)
 
 	struct egress_gw_policy_entry in_val_excluded_cidr = {
 		.egress_ip  = 0,
-		.gateway_ip = 0,
+		.gateway_ip = EGRESS_GATEWAY_EXCLUDED_CIDR,
 	};
 
 	map_update_elem(&EGRESS_POLICY_MAP, &in_key_excluded_cidr, &in_val_excluded_cidr, 0);

--- a/bpf/tests/tc_egressgw_redirect.c
+++ b/bpf/tests/tc_egressgw_redirect.c
@@ -118,7 +118,7 @@ int egressgw_redirect_setup(struct __ctx_buff *ctx)
 		.deny = 0,
 	};
 
-	/* avoid policy drop */
+	/* Avoid policy drop */
 	map_update_elem(&POLICY_MAP, &policy_key, &policy_value, BPF_ANY);
 
 	/* Jump into the entrypoint */
@@ -231,7 +231,7 @@ int egressgw_skip_excluded_cidr_redirect_setup(struct __ctx_buff *ctx)
 		.deny = 0,
 	};
 
-	/* avoid policy drop */
+	/* Avoid policy drop */
 	map_update_elem(&POLICY_MAP, &policy_key, &policy_value, BPF_ANY);
 
 	/* Jump into the entrypoint */
@@ -267,6 +267,117 @@ int egressgw_skip_excluded_cidr_redirect_check(const struct __ctx_buff *ctx)
 	};
 
 	map_delete_elem(&EGRESS_POLICY_MAP, &in_key_excluded_cidr);
+
+	test_finish();
+}
+
+/* Test that a packet matching an egress gateway policy without a gateway on the
+ * from-container program does not get redirected to the gateway node.
+ */
+PKTGEN("tc", "tc_egressgw_skip_no_gateway_redirect")
+int egressgw_skip_no_gateway_redirect_pktgen(struct __ctx_buff *ctx)
+{
+	struct pktgen builder;
+	struct tcphdr *l4;
+	struct ethhdr *l2;
+	struct iphdr *l3;
+	void *data;
+
+	/* Init packet builder */
+	pktgen__init(&builder, ctx);
+
+	/* Push ethernet header */
+	l2 = pktgen__push_ethhdr(&builder);
+	if (!l2)
+		return TEST_ERROR;
+
+	ethhdr__set_macs(l2, (__u8 *)client_mac, (__u8 *)ext_svc_mac);
+
+	/* Push IPv4 header */
+	l3 = pktgen__push_default_iphdr(&builder);
+	if (!l3)
+		return TEST_ERROR;
+
+	l3->saddr = CLIENT_IP;
+	l3->daddr = EXTERNAL_SVC_IP;
+
+	/* Push TCP header */
+	l4 = pktgen__push_default_tcphdr(&builder);
+	if (!l4)
+		return TEST_ERROR;
+
+	l4->source = CLIENT_PORT;
+	l4->dest = EXTERNAL_SVC_PORT;
+
+	data = pktgen__push_data(&builder, default_data, sizeof(default_data));
+	if (!data)
+		return TEST_ERROR;
+
+	/* Calc lengths, set protocol fields and calc checksums */
+	pktgen__finish(&builder);
+
+	return 0;
+}
+
+SETUP("tc", "tc_egressgw_skip_no_gateway_redirect")
+int egressgw_skip_no_gateway_redirect_setup(struct __ctx_buff *ctx)
+{
+	struct egress_gw_policy_key in_key_no_gateway = {
+		.lpm_key = { EGRESS_PREFIX_LEN(32), {} },
+		.saddr   = CLIENT_IP,
+		.daddr   = EXTERNAL_SVC_IP,
+	};
+
+	struct egress_gw_policy_entry in_val_no_gateway = {
+		.egress_ip  = 0,
+		.gateway_ip = EGRESS_GATEWAY_NO_GATEWAY,
+	};
+
+	map_update_elem(&EGRESS_POLICY_MAP, &in_key_no_gateway, &in_val_no_gateway, 0);
+
+	struct policy_key policy_key = {
+		.egress = 1,
+	};
+	struct policy_entry policy_value = {
+		.deny = 0,
+	};
+
+	/* Avoid policy drop */
+	map_update_elem(&POLICY_MAP, &policy_key, &policy_value, BPF_ANY);
+
+	/* Jump into the entrypoint */
+	tail_call_static(ctx, &entry_call_map, FROM_CONTAINER);
+	/* Fail if we didn't jump */
+	return TEST_ERROR;
+}
+
+CHECK("tc", "tc_egressgw_skip_no_gateway_redirect")
+int egressgw_skip_no_gateway_redirect_check(const struct __ctx_buff *ctx)
+{
+	void *data, *data_end;
+	__u32 *status_code;
+
+	test_init();
+
+	data = (void *)(long)ctx_data(ctx);
+	data_end = (void *)(long)ctx->data_end;
+
+	if (data + sizeof(__u32) > data_end)
+		test_fatal("status code out of bounds");
+
+	status_code = data;
+	assert(*status_code == CTX_ACT_OK);
+
+	/* Delete the no gateway entry otherwise other tests may fail as this
+	 * entry will persist across the different tests.
+	 */
+	struct egress_gw_policy_key in_key_no_gateway = {
+		.lpm_key = { EGRESS_PREFIX_LEN(32), {} },
+		.saddr   = CLIENT_IP,
+		.daddr   = EXTERNAL_SVC_IP,
+	};
+
+	map_delete_elem(&EGRESS_POLICY_MAP, &in_key_no_gateway);
 
 	test_finish();
 }

--- a/bpf/tests/tc_egressgw_snat.c
+++ b/bpf/tests/tc_egressgw_snat.c
@@ -256,7 +256,7 @@ int egressgw_skip_excluded_cidr_snat_setup(struct __ctx_buff *ctx)
 
 	struct egress_gw_policy_entry in_val_excluded_cidr = {
 		.egress_ip  = 0,
-		.gateway_ip = 0,
+		.gateway_ip = EGRESS_GATEWAY_EXCLUDED_CIDR,
 	};
 
 	map_update_elem(&EGRESS_POLICY_MAP, &in_key_excluded_cidr, &in_val_excluded_cidr, 0);

--- a/cilium/cmd/bpf_egress_list.go
+++ b/cilium/cmd/bpf_egress_list.go
@@ -7,6 +7,7 @@ import (
 	"errors"
 	"fmt"
 	"io/fs"
+	"net"
 	"os"
 	"text/tabwriter"
 
@@ -14,6 +15,7 @@ import (
 
 	"github.com/cilium/cilium/pkg/command"
 	"github.com/cilium/cilium/pkg/common"
+	"github.com/cilium/cilium/pkg/egressgateway"
 	"github.com/cilium/cilium/pkg/maps/egressmap"
 )
 
@@ -51,7 +53,7 @@ var bpfEgressListCmd = &cobra.Command{
 				SourceIP:  key.GetSourceIP().String(),
 				DestCIDR:  key.GetDestCIDR().String(),
 				EgressIP:  val.GetEgressIP().String(),
-				GatewayIP: val.GetGatewayIP().String(),
+				GatewayIP: mapGatewayIP(val.GetGatewayIP()),
 			})
 		}
 
@@ -72,6 +74,18 @@ var bpfEgressListCmd = &cobra.Command{
 			printEgressList(bpfEgressList)
 		}
 	},
+}
+
+// This function attempt to translate gatewayIP to special values if they exist
+// or return the IP as a string otherwise.
+func mapGatewayIP(ip net.IP) string {
+	if ip.Equal(egressgateway.GatewayNotFoundIPv4) {
+		return "Not Found"
+	}
+	if ip.Equal(egressgateway.ExcludedCIDRIPv4) {
+		return "Excluded CIDR"
+	}
+	return ip.String()
 }
 
 func printEgressList(egressList []egressPolicy) {

--- a/pkg/egressgateway/manager.go
+++ b/pkg/egressgateway/manager.go
@@ -31,8 +31,13 @@ import (
 )
 
 var (
-	log      = logging.DefaultLogger.WithField(logfields.LogSubsys, "egressgateway")
-	zeroIPv4 = net.ParseIP("0.0.0.0")
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "egressgateway")
+	// GatewayNotFoundIPv4 is a special IP value used as gatewayIP in the BPF policy
+	// map to indicate no gateway was found for the given policy
+	GatewayNotFoundIPv4 = net.ParseIP("0.0.0.0")
+	// ExcludedCIDRIPv4 is a special IP value used as gatewayIP in the BPF policy map
+	// to indicate the entry is for an excluded CIDR and should skip egress gateway
+	ExcludedCIDRIPv4 = net.ParseIP("0.0.0.1")
 )
 
 // Cell provides a [Manager] for consumption with hive.
@@ -530,7 +535,7 @@ func (manager *Manager) addMissingEgressRules() {
 
 		gatewayIP := gwc.gatewayIP
 		if excludedCIDR {
-			gatewayIP = zeroIPv4
+			gatewayIP = ExcludedCIDRIPv4
 		}
 
 		if policyPresent && policyVal.Match(gwc.egressIP.IP, gatewayIP) {
@@ -570,7 +575,7 @@ nextPolicyKey:
 		matchPolicy := func(endpointIP net.IP, dstCIDR *net.IPNet, excludedCIDR bool, gwc *gatewayConfig) bool {
 			gatewayIP := gwc.gatewayIP
 			if excludedCIDR {
-				gatewayIP = zeroIPv4
+				gatewayIP = ExcludedCIDRIPv4
 			}
 
 			return policyKey.Match(endpointIP, dstCIDR) && policyVal.Match(gwc.egressIP.IP, gatewayIP)

--- a/pkg/egressgateway/manager_privileged_test.go
+++ b/pkg/egressgateway/manager_privileged_test.go
@@ -56,6 +56,8 @@ const (
 	egressCIDR1 = "192.168.101.1/24"
 
 	zeroIP4 = "0.0.0.0"
+
+	gatewayExcludedCIDRValue = "0.0.0.1" // Special value for gatewayIP excluded CIDR, see pkg/egressgateway/manager.go
 )
 
 var (
@@ -245,7 +247,7 @@ func (k *EgressGatewayTestSuite) TestEgressGatewayManager(c *C) {
 
 	assertEgressRules(c, []egressRule{
 		{ep1IP, destCIDR, egressIP1, node1IP},
-		{ep1IP, excludedCIDR1, egressIP1, zeroIP4},
+		{ep1IP, excludedCIDR1, egressIP1, gatewayExcludedCIDRValue},
 		{ep2IP, destCIDR, zeroIP4, node2IP},
 	})
 
@@ -279,8 +281,8 @@ func (k *EgressGatewayTestSuite) TestEgressGatewayManager(c *C) {
 
 	assertEgressRules(c, []egressRule{
 		{ep1IP, destCIDR, egressIP1, node1IP},
-		{ep1IP, excludedCIDR1, egressIP1, zeroIP4},
-		{ep1IP, excludedCIDR2, egressIP1, zeroIP4},
+		{ep1IP, excludedCIDR1, egressIP1, gatewayExcludedCIDRValue},
+		{ep1IP, excludedCIDR2, egressIP1, gatewayExcludedCIDRValue},
 		{ep2IP, destCIDR, zeroIP4, node2IP},
 	})
 
@@ -320,7 +322,7 @@ func (k *EgressGatewayTestSuite) TestEgressGatewayManager(c *C) {
 
 	assertEgressRules(c, []egressRule{
 		{ep1IP, destCIDR, egressIP1, node1IP},
-		{ep1IP, excludedCIDR2, egressIP1, zeroIP4},
+		{ep1IP, excludedCIDR2, egressIP1, gatewayExcludedCIDRValue},
 		{ep2IP, destCIDR, zeroIP4, node2IP},
 	})
 

--- a/pkg/egressgateway/manager_privileged_test.go
+++ b/pkg/egressgateway/manager_privileged_test.go
@@ -57,7 +57,9 @@ const (
 
 	zeroIP4 = "0.0.0.0"
 
-	gatewayExcludedCIDRValue = "0.0.0.1" // Special value for gatewayIP excluded CIDR, see pkg/egressgateway/manager.go
+	// Special values for gatewayIP, see pkg/egressgateway/manager.go
+	gatewayNotFoundValue     = "0.0.0.0"
+	gatewayExcludedCIDRValue = "0.0.0.1"
 )
 
 var (
@@ -66,11 +68,13 @@ var (
 
 	identityAllocator = testidentity.NewMockIdentityAllocator(nil)
 
-	nodeGroup1Labels = map[string]string{"label1": "1"}
-	nodeGroup2Labels = map[string]string{"label2": "2"}
+	nodeGroupNotFoundLabels = map[string]string{"label1": "notfound"}
+	nodeGroup1Labels        = map[string]string{"label1": "1"}
+	nodeGroup2Labels        = map[string]string{"label2": "2"}
 
-	nodeGroup1Selector = &slimv1.LabelSelector{MatchLabels: nodeGroup1Labels}
-	nodeGroup2Selector = &slimv1.LabelSelector{MatchLabels: nodeGroup2Labels}
+	nodeGroupNotFoundSelector = &slimv1.LabelSelector{MatchLabels: nodeGroupNotFoundLabels}
+	nodeGroup1Selector        = &slimv1.LabelSelector{MatchLabels: nodeGroup1Labels}
+	nodeGroup2Selector        = &slimv1.LabelSelector{MatchLabels: nodeGroup2Labels}
 )
 
 type ipRule struct {
@@ -354,6 +358,17 @@ func (k *EgressGatewayTestSuite) TestEgressGatewayManager(c *C) {
 	assertIPRules(c, []ipRule{
 		{ep1IP, destCIDR, egressCIDR1, testInterface1Idx},
 	})
+
+	// Test matching no gateway
+	policy1 = newEgressPolicyConfigWithNodeSelector("policy-1", ep1Labels, destCIDR, []string{}, nodeGroupNotFoundSelector, testInterface1)
+	egressGatewayManager.OnAddEgressPolicy(policy1)
+
+	assertEgressRules(c, []egressRule{
+		{ep1IP, destCIDR, zeroIP4, gatewayNotFoundValue},
+		{ep2IP, destCIDR, zeroIP4, node2IP},
+	})
+
+	assertIPRules(c, []ipRule{})
 
 	// Update the endpoint labels in order for it to not be a match
 	_ = updateEndpointAndIdentity(&ep1, id1, map[string]string{})

--- a/pkg/egressgateway/policy.go
+++ b/pkg/egressgateway/policy.go
@@ -101,7 +101,7 @@ func (config *policyGatewayConfig) selectsNodeAsGateway(node nodeTypes.Node) boo
 func (config *PolicyConfig) regenerateGatewayConfig(manager *Manager) {
 	gwc := gatewayConfig{
 		egressIP:  net.IPNet{IP: net.IPv4zero, Mask: net.CIDRMask(0, 0)},
-		gatewayIP: zeroIPv4,
+		gatewayIP: GatewayNotFoundIPv4,
 	}
 
 	policyGwc := config.policyGwConfig

--- a/test/k8s/egress.go
+++ b/test/k8s/egress.go
@@ -407,6 +407,46 @@ var _ = SkipDescribeIf(func() bool {
 					})
 				})
 			})
+			Context("egress gw policy when the gateway is not found", func() {
+				var policyYAML string
+
+				BeforeAll(func() {
+					policyYAML = applyEgressPolicy("egress-gateway-policy-not-found.yaml")
+
+					// Wait for 2 entries:
+					// - 1 policy with 2 matching endpoints
+
+					err := kubectl.WaitForEgressPolicyEntries(helpers.K8s1, 2)
+					Expect(err).Should(BeNil(), "Failed waiting for egress policy map entries")
+
+					err = kubectl.WaitForEgressPolicyEntries(helpers.K8s2, 2)
+					Expect(err).Should(BeNil(), "Failed waiting for egress policy map entries")
+				})
+				AfterAll(func() {
+					kubectl.Delete(policyYAML)
+
+					err := kubectl.WaitForEgressPolicyEntries(helpers.K8s1, 0)
+					Expect(err).Should(BeNil(), "Failed waiting for egress policy map entries")
+
+					err = kubectl.WaitForEgressPolicyEntries(helpers.K8s2, 0)
+					Expect(err).Should(BeNil(), "Failed waiting for egress policy map entries")
+				})
+
+				AfterFailed(func() {
+					kubectl.CiliumReport("cilium bpf egress list", "cilium bpf nat list")
+				})
+
+				It("Traffic is not SNATed with egress gateway IP", func() {
+					testEgressGateway(&egressGatewayTestOpts{
+						fromGateway:    false,
+						shouldBeSNATed: false,
+					})
+					testEgressGateway(&egressGatewayTestOpts{
+						fromGateway:    true,
+						shouldBeSNATed: false,
+					})
+				})
+			})
 		})
 	}
 

--- a/test/k8s/manifests/egress-gateway-policy-not-found.yaml
+++ b/test/k8s/manifests/egress-gateway-policy-not-found.yaml
@@ -1,0 +1,18 @@
+apiVersion: cilium.io/v2
+kind: CiliumEgressGatewayPolicy
+metadata:
+  name: cegp-sample
+spec:
+  selectors:
+  - podSelector:
+      matchLabels:
+        zgroup: testDSClient
+    namespaceSelector:
+      matchLabels:
+        ns: cilium-test
+  destinationCIDRs:
+  - INPUT_OUTSIDE_NODE_IP/32
+  egressGateway:
+    nodeSelector:
+      matchLabels:
+        cilium.io/ci-node: not-found

